
<h2 style="text-align: center;">Desafío Equinox AI and Data Lab </h2>





Saludos aspirante:
Estás a punto de comenzar un reto para demostrar que puedes ser parte de Equinox.

Lee con atención las instrucciones para que puedas llevar a cabo tu prueba:

El Desafío Equinox **evaluará 4 aspectos básicos** que se requieren a la hora de elegir a un perfil. 
Por lo tanto la prueba estará dividida en **4 partes**.

- La primera parte evaluará tus conocimientos en **análisis y tratamiento de información**, esto junto a tu habilidad técnica en jupyter Notebooks, Python y demás librerias requeridas en Data analysis.

- La segunda parte esta orientada a conocer tu desempeño en la construcción de modelos de **Machine Learning**, Ten en cuenta seguir el pipeline completo para demostrar que tu modelo tiene un correcto desempeño.

- La tercera parte es un ejercicio orientado a evaluar tus **capacidades en SQL**. Consta de tres preguntas de negocio con la cuales queremos observar tu habilidad en el lenguaje.

- La cuarta parte es un ejercicio de **comunicación**, en donde serás entrevistado por líderes del área y tendrás que contarnos sobre el modo en que abordaste el desafío.

<h4>Reglas para el desafío</h4>

- El desafío podrá ser resuelto en aproximadamente 3 días, **comenzando el viernes 20 de Mayo a las 2:00 pm  y terminará el Lunes 23 de Mayo a las 01:00 pm** .

- Es necesario contar con una cuenta de github para poder descargar el contenido del desafío y subir tus respuestas. Si no tienes una cuenta puedes dirigirte al sitio https://github.com/ y crear una.

- Para la primera y segunda parte se utilizarán Jupyter Notebooks, Python y sus librerías orientadas a análisis de Datos y Machine Learning. Asegúrate de tener instaladas estas herramientas en tu equipo antes del lanzamiento del desafío para que puedas dedicar tus esfuerzos a resolver los problemas.

- La tercera parte tomará la base de datos que esta alojada en https://www.w3schools.com/sql/trysql.asp?filename=trysql_select_all como referencia para las preguntas planteadas. Allí encontrarás una pequeña base de datos y un sandbox donde podrás probar tus queries.

- Todo lo que construyas debe estár en un Jupyter Notebook, La primera y segunda parte deben correr dentro del notebook, mientras que para la tercera parte, solamente debes colocar los queries que resuelvan las preguntas de esta parte. **Asegurate de ser ordenado** para que nuestro proceso de validación pueda ser ágil.

- Asegúrate de dar fork al repositorio https://github.com/EquinoxAI/desafio_equinox con tu nombre y apellido. **Ten en cuenta siempre  dar commit y subir archivos, para poder evaluar tu prueba.**
![example1](img/fork1.png)
![example2](img/fork2.png)

- Debes estar atento las próximas semanas para realizar una entrevista, en la cuál deberás sustentar temas puntuales de tu prueba.


Desde Equinox te deseamos muchos éxitos en este gran desafío y esperamos que des lo mejor de tí.

Te esperamos.




<h4> Primera Parte: Análisis y tratamiento de información </h4>

- Tomando como referencia el archivo **kc_house_data.csv** responde las siguientes preguntas:

    - ¿Cuántos registros y columnas esta compuesto el dataset?
    - ¿Qué años componen el dataset y que cantidad de registros hay para cada año?
    - ¿Cuál es la media de habitaciones de las propiedades que conforman el dataset?
    - ¿Cuál es el precio promedio de alquilar una casa  de 3 pisos o más?
    - ¿Cuántos pisos posée la casa cuyo número de habitaciones es igual a 33?
    - ¿Cuántas propiedades son del año 2015 y tienen un precio mayor a 400000 dólares?
    - ¿Cuál es la distribución del número de pisos en el dataset?. Demuéstralo a través de un gráfico de barras ordenado
    - ¿Con que tipo de distribución relacionaría el precio de las propiedades cuyo valor es menor o igual a 2000000 dentro del dataset         Demuéstralo con un histograma.

 <h4>Lectura del archivo </h4>

In [20]:
import pandas as pd
path= 'kc_house_data.csv'
df=pd.read_csv(path)
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

<h4>¿Cuántos registros y columnas esta compuesto el dataset? </h4>

In [21]:
df.shape

(21613, 21)

Hay 21613 registros y 21 colúmnas

<h4>¿Qué años componen el dataset y que cantidad de registros hay para cada año? </h4>

Para la variable yr_built

In [22]:
yr_built_counts = df['yr_built'].value_counts().to_frame()
yr_built_counts.rename(columns={'yr_built': 'value_counts'}, inplace=True)
yr_built_counts

value_counts
2014           559
2006           454
2005           450
2004           433
2003           422
...            ...
1933            30
1901            29
1902            27
1935            24
1934            21

[116 rows x 1 columns]

Para la variable yr_renovated

In [23]:
yr_renovated_counts = df['yr_renovated'].value_counts().to_frame()
yr_renovated_counts.rename(columns={'yr_renovated': 'value_counts'}, inplace=True)
yr_renovated_counts

value_counts
0            20699
2014            91
2013            37
2003            36
2000            35
...            ...
1934             1
1959             1
1951             1
1948             1
1944             1

[70 rows x 1 columns]

<h4>¿Cuál es la media de habitaciones de las propiedades que conforman el dataset?</h4>

In [24]:
df['bedrooms'].mean()

3.37084162309721

La media es 3.37084162309721 habitaciones por propiedad

<h4>¿Cuál es el precio promedio de alquilar una casa  de 3 pisos o más? </h4>

<h4>¿Cuántos pisos posée la casa cuyo número de habitaciones es igual a 33? </h4>

<h4>¿Cuántas propiedades son del año 2015 y tienen un precio mayor a 400000 dólares? </h4>

<h4>¿Cuál es la distribución del número de pisos en el dataset?. Demuéstralo a través de un gráfico de barras ordenado </h4>

<h4>¿Con que tipo de distribución relacionaría el precio de las propiedades cuyo valor es menor o igual a 2000000 dentro del dataset         Demuéstralo con un histograma. </h4>

In [25]:
df.dtypes

id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

In [26]:
df.describe(include='all')

id             date         price      bedrooms  \
count   2.161300e+04            21613  2.161300e+04  21613.000000   
unique           NaN              372           NaN           NaN   
top              NaN  20140623T000000           NaN           NaN   
freq             NaN              142           NaN           NaN   
mean    4.580302e+09              NaN  5.400881e+05      3.370842   
std     2.876566e+09              NaN  3.671272e+05      0.930062   
min     1.000102e+06              NaN  7.500000e+04      0.000000   
25%     2.123049e+09              NaN  3.219500e+05      3.000000   
50%     3.904930e+09              NaN  4.500000e+05      3.000000   
75%     7.308900e+09              NaN  6.450000e+05      4.000000   
max     9.900000e+09              NaN  7.700000e+06     33.000000   

           bathrooms   sqft_living      sqft_lot        floors    waterfront  \
count   21613.000000  21613.000000  2.161300e+04  21613.000000  21613.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        2.114757   2079.899736  1.510697e+04      1.494309      0.007542   
std         0.770163    918.440897  4.142051e+04      0.539989      0.086517   
min         0.000000    290.000000  5.200000e+02      1.000000      0.000000   
25%         1.750000   1427.000000  5.040000e+03      1.000000      0.000000   
50%         2.250000   1910.000000  7.618000e+03      1.500000      0.000000   
75%         2.500000   2550.000000  1.068800e+04      2.000000      0.000000   
max         8.000000  13540.000000  1.651359e+06      3.500000      1.000000   

                view  ...         grade    sqft_above  sqft_basement  \
count   21613.000000  ...  21613.000000  21613.000000   21613.000000   
unique           NaN  ...           NaN           NaN            NaN   
top              NaN  ...           NaN           NaN            NaN   
freq             NaN  ...           NaN           NaN            NaN   
mean        0.234303  ...      7.656873   1788.390691     291.509045   
std         0.766318  ...      1.175459    828.090978     442.575043   
min         0.000000  ...      1.000000    290.000000       0.000000   
25%         0.000000  ...      7.000000   1190.000000       0.000000   
50%         0.000000  ...      7.000000   1560.000000       0.000000   
75%         0.000000  ...      8.000000   2210.000000     560.000000   
max         4.000000  ...     13.000000   9410.000000    4820.000000   

            yr_built  yr_renovated       zipcode           lat          long  \
count   21613.000000  21613.000000  21613.000000  21613.000000  21613.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean     1971.005136     84.402258  98077.939805     47.560053   -122.213896   
std        29.373411    401.679240     53.505026      0.138564      0.140828   
min      1900.000000      0.000000  98001.000000     47.155900   -122.519000   
25%      1951.000000      0.000000  98033.000000     47.471000   -122.328000   
50%      1975.000000      0.000000  98065.000000     47.571800   -122.230000   
75%      1997.000000      0.000000  98118.000000     47.678000   -122.125000   
max      2015.000000   2015.000000  98199.000000     47.777600   -121.315000   

        sqft_living15     sqft_lot15  
count    21613.000000   21613.000000  
unique            NaN            NaN  
top               NaN            NaN  
freq              NaN            NaN  
mean      1986.552492   12768.455652  
std        685.391304   27304.179631  
min        399.000000     651.000000  
25%       1490.000000    5100.000000  
50%       1840.000000    7620.000000  
75%       23

<h4> Segunda Parte: Machine Learning </h4>

Tomando el dataset de los ejercicios anteriores entrena un modelo que use inteligencia artificial que prediga el precio de propiedades futuras.

Ten en cuenta: 
- Plantear una correcta división del dataset para entrenar y testear. 
- Usar el modelo correcto.
- Métricas de desemempeño.
- Construir una función para poder inferir resultados fuera del dataset.

<h4> Tercera Parte: Capacidades en SQL </h4>

Resuelve las siguientes preguntas a través de una consulta SQL, La base de datos que usaremos como referencia, se encuentra en:
https://www.w3schools.com/sql/trysql.asp?filename=trysql_select_all
Allí también podrás encontrar un sandbox para realizar las consultas de una manera sencilla.


- Cuales son las categorias de productos que mas han vendido y cuanto han vendido en precio y cantidad.
- Cuantas unidades y dinero de cada producto ha vendido cada empleado 
- Cual es el producto más vendido por cada empleado. 

El orden,buenas prácticas y presentación de las consultas también cuenta. Asegurate de colocarlas en el notebook para que las podamos revisar.